File: smips/python/smips/ProblemData.py
from __future__ import annotations

from utils import CppClass, get_library

_LIB = get_library()


class ProblemData(CppClass):

    def __init__(self,
                 A,
                 T,
                 W,
                 scenarios,
                 scenario_probabilities,
                 first_stage_constr_senses,
                 second_stage_constr_senses,
                 first_stage_var_types,
                 second_stage_var_types,
                 first_stage_coeffs,
                 second_stage_coeffs,
                 first_stage_lower_bound,
                 first_stage_upper_bound,
                 second_stage_lower_bound,
                 second_stage_upper_bound,
                 first_stage_rhs):
        pass

    @classmethod
    def fromSmps(cls, location: str) -> ProblemData:
        pass

    def firstStageConstrSenses(self):
        pass

    def secondStageConstrSenses(self):
        pass

    def firstStageVarTypes(self):
        pass

    def secondStageVarTypes(self):
        pass

    def nScenarios(self):
        pass

    def Amat(self):
        pass

    def Tmat(self):
        pass

    def Wmat(self):
        pass

    def firstStageCoeffs(self):
        pass

    def secondStageCoeffs(self):
        pass

    def firstStageRhs(self):
        pass

    def firstStageLowerBound(self):
        pass

    def firstStageUpperBound(self):
        pass

    def secondStageLowerBound(self):
        pass

    def secondStageUpperBound(self):
        pass

    def scenarios(self):
        pass

    def probability(self, scenario: int) -> float:
        pass

File: lbda/src/problemdata.cpp
#include "problemdata.h"

#include "smps/smps.h"

#include <utility>

ProblemData::ProblemData(arma::sp_mat Amat,
                         arma::sp_mat Tmat,
                         arma::sp_mat Wmat,
                         arma::mat scenarios,
                         arma::vec scenarioProbabilities,
                         arma::Col<char> firstStageConstrSenses,
                         arma::Col<char> secondStageConstrSenses,
                         arma::Col<char> firstStageVarTypes,
                         arma::Col<char> secondStageVarTypes,
                         arma::vec firstStageCoeffs,
                         arma::vec secondStageCoeffs,
                         arma::vec firstStageLowerBound,
                         arma::vec firstStageUpperBound,
                         arma::vec secondStageLowerBound,
                         arma::vec secondStageUpperBound,
                         arma::vec firstStageRhs) :
    d_Amat(std::move(Amat)),
    d_Tmat(std::move(Tmat)),
    d_Wmat(std::move(Wmat)),
    d_scenarios(std::move(scenarios)),
    d_scenarioProbabilities(std::move(scenarioProbabilities)),
    d_firstStageConstrSenses(std::move(firstStageConstrSenses)),
    d_secondStageConstrSenses(std::move(secondStageConstrSenses)),
    d_firstStageVarTypes(std::move(firstStageVarTypes)),
    d_secondStageVarTypes(std::move(secondStageVarTypes)),
    d_firstStageCoeffs(std::move(firstStageCoeffs)),
    d_secondStageCoeffs(std::move(secondStageCoeffs)),
    d_firstStageLowerBound(std::move(firstStageLowerBound)),
    d_firstStageUpperBound(std::move(firstStageUpperBound)),
    d_secondStageLowerBound(std::move(secondStageLowerBound)),
    d_secondStageUpperBound(std::move(secondStageUpperBound)),
    d_firstStageRhs(std::move(firstStageRhs))
{
}

ProblemData ProblemData::fromSmps(std::string const &location)
{
    smps::Smps smps;
    smps.read(location);

    return ProblemData{smps.Amat(),
                       smps.Tmat(),
                       smps.Wmat(),
                       smps.generateScenarios(),
                       smps.scenarioProbabilities(),
                       smps.firstStageConstrSenses(),
                       smps.secondStageConstrSenses(),
                       smps.firstStageVarTypes(),
                       smps.secondStageVarTypes(),
                       smps.firstStageObjCoeffs(),
                       smps.secondStageObjCoeffs(),
                       smps.firstStageLowerBound(),
                       smps.firstStageUpperBound(),
                       smps.secondStageLowerBound(),
                       smps.secondStageUpperBound(),
                       smps.firstStageRhs()};
}
